### ISSUE TITLE
Using desktop background can be offset with multiple monitors

### ISSUE THREAD

**Issue by jonathonmckay** [mintty/mintty#1296]
After years of using just the transparency setting found in Options, I have recently discovered the command line option to set a background image. Of particular interest is the mode where it grabs the desktop background to make it look like the window is hardly there at all, like this:
`echo -ne '\e]11;=,99\a'`
However, I have found that with my multi monitor setup, unless that "primary monitor" is the left most monitor (which it's not since I want the center monitor to be "primary"), and if there is a separate image on each monitor, the image shown on the background is from the mintty window, but from the monitor one position to the left.
Like so:
![image](https://github.com/user-attachments/assets/c40d88eb-d3b7-43fb-9f4b-e2249f88571b)
I suspect this is because the left monitor would have a negative location relative to the primary monitor.
I tried setting the left most monitor as "primary", and then it works properly, but then new windows and notifications don't pop up front and center which interferes with my workflow.

**Comment by mintty** [comment]
I'm not sure how to interpret your screenshot. The mintty window has a greyish background, not a desktop pattern.
Also note that desktop background only works with tiled desktop, not full-size image.
Then, in order to test this properly, I'd need information on how to setup different backgrounds for two monitors, which is among the information well hidden by MS.

**Comment by jonathonmckay** [comment]
Apologies.
The image on the other monitor at the time I took the screen shot did involve quite a bit of flat grey.
I'll make a new screen shot:
![image](https://github.com/user-attachments/assets/ec57d660-7973-490a-a161-def19721d94b)

The image inside the mintty window does move around as I move it around the monitor, just off set by one monitor.

To make different image on each monitor, I'm using a program called John's Background Switcher (https://johnsad.ventures/software/backgroundswitcher/windows/). It takes three random images from folders you specify an concatenates them into one large image, which it then (somehow) tells Windows to apply as the background image. It also automatically sets the "Fit" to "Tile" (but the image is perfectly scaled to your entire display space, so it only shows once).

A likely easy way to solve this would be a setting in some config file to set a windows background image X and Y offset (I say this admittedly not knowing anything about your code). Someone with this rare problem could then just set the X offset to -1920 (or whatever). Just an idea.

**Comment by mintty** [comment]
I've reproduced the issue. I had hoped for an esay fix, just tweaking the image offset when grabbing the background image and using it as a "brush". But there's no offset handling currently in effect (function load_background_brush) so it could be more tricky...

**Comment by mintty** [comment]
... except maybe in the case of using GDI+ (function load_background_image_brush) but still tricky as that's a complex function...

**Comment by mintty** [comment]
... or rather, there's function offset_bg already, to adjust to the window position...

**Comment by mintty** [comment]
I fixed the issue and also implemented dynamic updating while the window is moved or resized. Testing appreciated.

**Comment by jonathonmckay** [comment]
I pulled the newest code and built it.
It works as expected. This is exactly what I was hoping for.
![image](https://github.com/user-attachments/assets/b24d728e-c8bd-440d-a351-1cf425176ba8)

I have made some observations:
- While moving the window, the window does lag and stutter some as the window moves.  Given the amount graphical redraw being done, this is understandable and acceptable.
- While sizing the window, there is lag and stutter, same as with moving it.
- If the background changes, the image inside does not update until next time the window is moved, sized, or gets focus.
- If the window is currently showing output from ASCII art generators (like TOIlet), the windows stutter is made much worse.
None of these are show stoppers; just things I've seen while checking it out.

Thank you so much

**Comment by mintty** [comment]
Thanks for testing and feedback.

> If the background changes, the image inside does not update until next time the window is moved, sized, or gets focus.

I don't see that. Quite the opposite, when I change the desktop background, the image changes within the mintty window even before it changes on the desktop.

> If the window is currently showing output from ASCII art generators (like TOIlet), the windows stutter is made much worse.

Is that coloured ASCII art, so with a lot of different attributes maybe? Can you upload a test file with it?

**Comment by jonathonmckay** [comment]
> when I change the desktop background, the image changes within the mintty window even before it changes on the desktop.

Isn't that peculiar. Wonder if its a Windows version thing? I am running Windows10, 22H2.
Also, when I tested Mintty above, it wasn't "installed", I was just running it with `./mintty.exe` from my installed Cygwin Mintty terminal in the compiled bin folder.
For me, either getting John's Background Switcher (5.6.0.5) to change the background, or changing the background by hand with the Windows personalization setting causes the windows desktop to change, but not the terminal background.

> Is that coloured ASCII art, so with a lot of different attributes maybe?

Um, maybe?
TOIlet doesn't seem to be available in Cygwin directly, rather its part of a login banner on a Linux server I use.
This command will cause the stutter very badly:
`toilet "TestTestTest" --font mono9 --termwidth`
While this command causes almost none:
`toilet "TestTestTest" --termwidth`
The second command uses more standard characters, while the first one seems to use interesting blocks to make it look "better". Maybe high Unicode characters? Not sure.
But, like I said, it's just the login banner, so once it scrolls of the top of the screen, the issue is basically gone.

**Comment by BrianInglis** [comment]
[Repology toilet](https://repology.org/projects/?search=toilet): free replacement for `figlet` by author(s) of and using `libcaca` color ASCII art library, which are both packaged for *Cygwin*. 
You could probably use `figlet` for testing. 
If needed I could run my conversion script from the `toilet` *Fedora* `spec` to *Cygwin* package `cygport` script, to build a package.

**Comment by mintty** [comment]
Thanks, and I did not know repology before.
Apparently figlet is restricted to ASCII, not my tool.

**Comment by mintty** [comment]
You could try to add one invocation behind WM_WININICHANGE:
```
    when WM_THEMECHANGED or WM_WININICHANGE or WM_SYSCOLORCHANGE:
        win_invalidate_all(true);
```
but I doubt that would make a change.

With toilet --font mono9, Unicode characters from the Block Elements range are used. This increases repaint flickering slightly, probably because those characters are self-drawn now in mintty, not taken from the font.

**Comment by jonathonmckay** [comment]
> You could try to add one invocation behind WM_WININICHANGE:

Actually, that makes it work.*
I added `win_invalidate_all(true);` on line 4748 of winmain.c, right after the comment block for the long `when`, and right before `win_adjust_borders`

(* This makes it work for background changes invoked by John's Background Switcher, but not background changes invoked by the Windows personalize settings. Weird.)

**Comment by mintty** [comment]
Weird indeed. As a last attempt, try to hook the refresh to another Windows event:
#define WM_DWMCOLORIZATIONCOLORCHANGED 0x0320
    when WM_DWMCOLORIZATIONCOLORCHANGED:
      win_invalidate_all(true);

**Comment by jonathonmckay** [comment]
(Kinda pushing what you're asking a non-C programmer to test, but I figured it out (I think))
Added the following code to winmain.c, at line 4740:

```
#ifndef WM_DWMCOLORIZATIONCOLORCHANGED 
#define WM_DWMCOLORIZATIONCOLORCHANGED  0x0320
#endif
    when WM_DWMCOLORIZATIONCOLORCHANGED:
      win_invalidate_all(true);
```

This did nothing as far as I can tell.

**Comment by jonathonmckay** [comment]
I've pulled that change you just committed. Works the same as my last test:

- Image offset is correct
- Image moves as window moves
- Background updates when Windows desktop updates (only when invoked by John's Background Switcher still, not Windows Personalization)

**Comment by jonathonmckay** [comment]
I've replaced the Cygwin "installed" version with the custom one I compiled for testing so I could easily daily drive it. 
I've found another small issue with the background.
If the window mode isn't set to normal (set to "frame" or "void") there is a slight offset vertically in the background image the height of the not-showing titlebar:

![image](https://github.com/user-attachments/assets/8c1e9d22-67eb-4d9d-92a5-46f62e3457cf)

**Comment by mintty** [comment]
Thanks for the testing. Uploaded a fix.

**Comment by jonathonmckay** [comment]
Recompiled and tested again.
Looks like the offsets are all gone:
![image](https://github.com/user-attachments/assets/d2760e8e-e020-4073-8469-b6f9df66dbca)
This image shows my background, the background showing through a "normal mintty window" (cyan border, right side), and a "void mintty top screen quake console" (no border, top of image). I don't see any offset now.

**Comment by mintty** [comment]
Thanks for the feedback. I was trying, again, to drop the limitation that the desktop background must be in "tiled" mode, but the proper scaling is too tricky for a quick solution.

**Comment by mintty** [comment]
Released 3.7.7.

**Comment by jonathonmckay** [comment]
Installed 3.7.7 via Cygwin.
Working as expected.
Thank you so much.
